include/deal.II/fe/fe_coupling_values.h

```diff
@@ -0,0 +1,653 @@
+  {
+    return prefix + "_" + Utilities::int_to_string(size) + "_" +
+           Utilities::type_to_string(exemplar_number);
+  }
+
+
+
+  template <typename ViewType>
+  typename RenumberedView<ViewType>::value_type
+  RenumberedView<ViewType>::value(const unsigned int shape_function,
+                                  const unsigned int q_point) const
+  {
+    AssertIndexRange(shape_function, data.n_dofs);
+    AssertIndexRange(q_point, data.n_quadrature_points);
+
+    const auto inner_shape_function = data.dof_renumbering.empty() ?
+                                        shape_function :
+                                        data.dof_renumbering[shape_function];
+    const auto inner_q_point        = data.quadrature_renumbering.empty() ?
+                                        q_point :
+                                        data.quadrature_renumbering[q_point];
+    if (inner_shape_function == numbers::invalid_unsigned_int)
+      return value_type(0);
+    else
+      {
+        AssertIndexRange(inner_shape_function, data.n_inner_dofs);
+        AssertIndexRange(inner_q_point, data.n_inner_quadrature_points);
+        return view.value(inner_shape_function, inner_q_point);
+      }
+  }
+
+
+
+  template <typename ViewType>
+  typename RenumberedView<ViewType>::gradient_type
+  RenumberedView<ViewType>::gradient(const unsigned int shape_function,
+                                     const unsigned int q_point) const
+  {
+    AssertIndexRange(shape_function, data.n_dofs);
+    AssertIndexRange(q_point, data.n_quadrature_points);
+
+    const auto inner_shape_function = data.dof_renumbering.empty() ?
+                                        shape_function :
+                                        data.dof_renumbering[shape_function];
+    const auto inner_q_point        = data.quadrature_renumbering.empty() ?
+                                        q_point :
+                                        data.quadrature_renumbering[q_point];
+    if (inner_shape_function == numbers::invalid_unsigned_int)
+      return gradient_type();
+    else
+      return view.gradient(inner_shape_function, inner_q_point);
+  }
+
+
+
+  template <typename ViewType>
+  template <typename ValueType>
+  std::vector<ValueType> &
+  RenumberedView<ViewType>::outer_to_inner_values(
+    std::vector<ValueType> &outer_values) const
+  {
+    AssertDimension(outer_values.size(), data.n_quadrature_points);
+    if (data.quadrature_renumbering.empty())
+      {
+        return outer_values;
+      }
+    else
+      {
+        const auto name = data.get_unique_container_name(
+          "RenumberedView::outer_to_inner_values",
+          data.n_inner_quadrature_points,
+          outer_values[0]);
+        auto &inner_values =
+          data.data_storage.get()
+            .template get_or_add_object_with_name<std::vector<ValueType>>(
+              name, data.n_inner_quadrature_points);
+        return inner_values;
```

Comment:
This pertains to the comment above regarding the mutex: We generally consider views themselves as stateless; the state is in the object being viewed. As a consequence using views is thread-safe is the underlying container is accessed in a thread-safe manner (for example because you're only ever reading).

However, here you have a state that you store in the `data` object. You need to guard all accesses to the `data.data_storage` access with a mutex and because `data` is a reference to another object, that mutex must be part of the `data` object, not the view object.

Reply:
Separately, I don't yet understand what you use this function for, but don't you have to make sure that `inner_values` actually has some content before you return it? What `get_or_add_object_with_name()` returns here is a vector with `data.n_inner_quadrature_points` elements, but no content. Is that what you want/need?

Reply:
The content is updated later. This method is used to get a reference to a vector of values (either the same as the input vector, if they have the same dimension and no renumbering must occur), which is then filled, i.e.,
```
auto &inner_values = outer_to_inner_values(values);
view.get_function_values(fe_function, inner_values);
inner_to_outer_values(inner_values, values);
```
`inner_values` is filled by `get_function_values`, and then reordered by `inner_to_outer_values`.

Reply:
I think I understood that later: You're just using this as a temp vector, but don't want to allocate/deallocate memory every time, right?